### PR TITLE
Update storefront events

### PR DESCRIPTION
**Pull request by cgebken**
Related task: https://epages.atlassian.net/browse/UC-3284

This PR is only for the NOW docs. The PR for the Beyond docs (without the wishlist event) will be a follow-up task in a different repo :)

### CODE REVIEW

File: apps/events.md [hunk 1]
```diff
@@ -5,21 +5,24 @@ title: Events
...
-| `product` | immutable.js object | Is triggered when the customer selects a product. Informs about the respective product as well as the current state of the cart. |
-| `category` | immutable.js object | Is triggered when the customer selects a category. Informs about the respective category and related products. |
-| `cart` | object | Is triggered when the customer accesses the cart. Informs about the current state of the cart, e.g. included items. |
-| `cart:add` | object | Is triggered when the customer adds a product to the cart. Provides information about the current state of the cart and the product that was added to it. |
-| `cart:setQuantity` | object | Is triggered when the quantity of a product in the cart is changed. This also includes the removal of a product. Informs about the quantity change and the affected product. |
-| `order:completed` | object | Is triggered when the customer reaches the order confirmation page after completing a purchase. Provides additional information about the order, such as the billing address, the order number, and the selected shipping method. |
+| Event | Description |
```
Comment (cgebken) [inline apps/events.md:14]: I've decided to remove the type column as I do not think that it has much value in this condensed version. The long version like on https://github.com/ePages-de/epages-ui/commit/cfa86e498 would be way to long. As developers can see the type in the example below, they can still get that information :)

File: apps/events.md [hunk 2]
```diff
@@ -166,6 +270,40 @@ updatedAt: "2018-04-17T14:48:29.000Z"
...
+  size: 5
+  __altered: false
+  __hash: undefined
+  __ownerID: undefined
+  _capacity: 5
+  _level: 5
+  _origin: 0
+  _root: null
+  _tail: Tt {array: Array(5), ownerID: undefined}
+}
```
Comment (bulutfatih) [inline apps/events.md:301]: I think this is wrong. It shows ImmutableJS object which is not useful. We should show plain javascript object instead. Am I correct @ingmarh ?

```suggestion
product: {
    availabilityText: "Available"
    available: true
    basePrice: {refQuantity: {…}, refPrice: {…}, formatted: "1 m³ = €0,12", quantity: {…}}
    bulkPrices: null
    conditionMicrodata: "NewCondition"
    customAttributes: [{…}]
    deliveryPeriod: "2-3"
    deliveryPeriodUnit: "DAYS"
    description: null
    energyLabel: null
    energyLabelSourceFile: null
    gtin: 7501054530107
    hasCrossSelling: false
    hasStockLevel: true
    hasVariations: false
    href: "/p/homemade-cherry-jam"
    image: null
    isVariationMaster: false
    isVariationProduct: false
    isVisible: true
    links: (5) [{…}, {…}, {…}, {…}, {…}]
    listPrice: null
    lowestPrice: null
    mainCategoryId: "5954B711-E377-2A90-C400-D5809AB3F62B"
    manufacturer: null
    manufacturerPrice: null
    metaDescription: ""
    name: "Homemade Cherry Jam"
    onStock: true
    outOfStock: false
    pickupPeriod: ""
    price: {taxType: "NET", formatted: "€40,00", amount: 40, currency: "EUR"}
    productDataSheet: null
    productId: "5954B706-E701-F357-A52D-D5809AB3F606"
    productVariationSelection: null
    productVariationValues: ""
    sku: "1007"
    slideshow: [{…}, {…}, {…}, {…}]
    slug: "homemade-cherry-jam"
    stockLevelClass: "in"
    stockLevelMicrodata: "InStock"
    title: "Homemade Cherry Jam"
    variationMaster: null
    variations: null
    vatNote: "components.productComponent.priceExclusiveVat"
    warnStock: false
}
```

File: apps/events.md (hunk 1 above)
Comment (bulutfatih) [reply apps/events.md:14]: Good point. I'm not sure if each developer has access to https://github.com/ePages-de/epages-ui 🤔

**Review by bulutfatih — CHANGES_REQUESTED** [review]

File: apps/events.md (hunk 1 above)
Comment (cgebken) [reply apps/events.md:14]: @bulutfatih For sure, they do not have access. But my point is that we have two options:

A: Extend the type information and copy the information that is stated in epages-ui in order to have a meaningful content for that column.

B: Remove the type column as we have the "You'll receive the following information:" part for each event. Developers can thus see the type themselves if they are interested in it (without the need to access epages-ui)

As I didn't want to have the big type column, I went with option B :)

File: apps/events.md (hunk 2 above)
Comment (cgebken) [reply apps/events.md:301]: @bulutfatih I'd also prefer the result you've proposed as it's more easy to understand, but what I included in this PR is what I got when having triggered the event myself... Did you get the another result? Do we need to edit the snippet developers should add to the code in order to get your proposal?

File: apps/events.md [hunk 3]
```diff
@@ -131,6 +135,106 @@ cart: {
...
+    stockLevelMicrodata: "InStock"
+    title: "Homemade Cherry Jam"
+    variationMaster: null
+    variations: null
+    vatNote: "components.productComponent.priceExclusiveVat"
+    warnStock: false
+}
+productIndex: 1
+type: "search"
+```
```
Comment (ingmarh) [inline apps/events.md:210]: I‘m not sure if it is good to have these kind of examples for each event, especially because the `console.log` example pattern is different or follows a different style throughout the examples.

The question that comes to my mind is: Is it important to list an example of the complete event details?

If it is enough to show _how to subscribe to these events_ and to tell the reader/app developer that they can _inspect/look at `event.detail`_ when they try out or develop their app, then only one example for all events could maybe also work.

Something like this (of course wording could be completely different):

```md
## Subscribing to an event

You can use the following code to subscribe to any event that is listed above.
Available information for each event can be found at the `event` property
"detail" in the event handler.

In this example, `event.detail` are output to the web browser‘s console.
Replace `EVENT_NAME` with the event you want to subscribe to, for example
"product".
```

```js
if (window.eComEventTarget) {
  window.eComEventTarget.addEventListener('EVENT_NAME', function (event) {
    // Do something with the event. All event information is in `event.detail`.
    console.log(event.detail)
  })
}
```

---

A problem is that for some events, the information is of type `Immutable.Map` that not every developer will know about, whereas for other events the information is in plain JavaScript object form. This mixture is not ideal in the first place (better would be if it is consistent). Issues I can think of:

- It _could_ lead to problems once real apps are using this storefront event API. Because if we change the types later on then it could be a breaking change for these apps. So a consistent format from the start would be best. Maybe we should make it consistent rather soon? So that we don‘t need to explain `Immutable.Map` or maintain it in the future.
- Now that there are plain JavaScript as well as `Immutable.Map` objects throughout the events, maybe we should document it on this page. [The "Type" information from the epages-ui document](https://github.com/ePages-de/epages-ui/blob/78c6cca10c5ae740f7719f942b0113ff556ace03/docs/javascript_events.md) that was now omitted on the table on this page for a good reason could give a hint. Still, that doesn‘t tell the developer what `Immutable.Map`s are or how to use them (but one could look it up). Also, for this, the examples for each type of event could help (which would speak against having only one example that documents how to subscribe to the events).

---

In any case, I think it would be good to at least make the examples consistent in their format (`console.log`). I could help with that. Maybe it would be easier to have a call about these topics if you agree with some of these points.

File: apps/events.md (hunk 2 above)
Comment (ingmarh) [reply apps/events.md:301]: I think this also has to do with confusion of different types of objects that one gets depending on which event they are subscribing. Sometimes it is `Immutable.Map`, sometimes a plain JavaScript object. Also just wrote about it in [this other annotation](https://github.com/ePages-de/epages-docs/pull/1190/files#r691347438).

I‘m not sure how to best proceed but I agree with @bulutfatih that if we list the example information here, then it is better to have it shown as plain JavaScript object.

To get a better idea what difference I‘m talking about:

```js
// This is how you access a property on a plain JavaScript object (where `event.detail.product` is the object):
event.detail.product.productId

// This is how you access a property on an Immutable.Map object (where `event.detail.product` is Immutable.Map):
event.detail.product.get('productId')

// And this is how you turn an Immutable.Map object to a plain JavaScript object:
event.detail.product.toJS()

// ... then you can just access "productId" like this:
event.detail.product.toJS().productId
```

I think it is confusing to have both Immutable.Map and plain JavaScript objects on the API that is now documented for app developers.

Comment (bulutfatih) [reply apps/events.md:301]: @ingmarh Does it affect the performance much if we turn `detail`_(immutable)_ to plain js object before dispatching? I'm afraid, yes. :/

I agree that we should find a consistence solution before it is late.

@cgebken no, you are right. I also get result with you. I just wanted to mention that result is not useful. Probably we will talk about this in tomorrow's JS Daily.

Comment (ingmarh) [reply apps/events.md:301]: Maybe using `.toJS()` before dispatching is okay for now. However, the API already exists. If there is an app that is already using it, it could be a breaking change. We can talk about it in tomorrow‘s JS Daily.

File: apps/events.md [hunk 4]
```diff
@@ -127,105 +150,448 @@ cart: {
...
-To make use of this event, you need to add the following snippet to your code:
+To get more information about this event, you can use the following snippet:
 
 ```js
 if (window.eComEventTarget) {
   window.eComEventTarget.addEventListener('cart:add', function (event) {
-    console.log('new cart data:',event.detail.cart,'added product:',event.detail.product,'quantity:',event.detail.quantity)
+    console.log('cart:add', {
+      new_cart_data: event.detail.cart,
+      added_product: event.detail.product,
```
Comment (ingmarh) [inline apps/events.md:584]: _Optional_:

These are descriptive names. However, given all the other examples I‘d prefer to name the variables the same as the property on `event.detail` for consistency. E.g. for `event.detail.cart`, it is "cart".

```suggestion
      cart: event.detail.cart,
      product: event.detail.product,
```

Or if you keep the naming, I suggest to make them camel case, which is more common in JavaScript. E.g. `newCartData`.

Comment (ingmarh) [inline apps/events.md:584]: Optional: These are descriptive names. However, given all the other examples I‘d prefer to name the variables the same as the property on `event.detail` for consistency. E.g. for `event.detail.cart`, it is "cart".

```suggestion
      cart: event.detail.cart,
      product: event.detail.product,
```

Or if you keep the naming, I suggest to make them camel case, which is more common in JavaScript. E.g. `newCartData`.

File: apps/events.md [hunk 5]
```diff
@@ -127,105 +150,448 @@ cart: {
...
   })
 }
 ```
 
-You'll receive the following information:
+Here's an example of what you can get:
 
 ```
-new cart data:{
+new_cart_data:{
```
Comment (ingmarh) [inline apps/events.md:594]: Minor: I suggest to add spacing for consistency with the other examples. Please also check the other examples.

```suggestion
new_cart_data: {
```

File: apps/events.md [hunk 6]
```diff
@@ -127,105 +150,448 @@ cart: {
...
+      lineItemId: event.detail.lineItemId,
+    })
+  })
+}
+```
+
+Here's an example of what you can get:
+
+```
+lineItemId: undefined
```
Comment (ingmarh) [inline apps/events.md:253]: Can we use a better example, or was it always `undefined` when you tried?

**Review by ingmarh — CHANGES_REQUESTED** [review]

File: apps/events.md [hunk 7]
```diff
@@ -341,19 +715,25 @@ quantity: 2
 quantityDelta: 1
 ```
 
-## Order:completed event
+### order:completed
+
+{% callout info Note: %}
+Please note that this event can only be triggered if the customer accepted all cookies in the cookie notice.
+{% endcallout %}
```
Comment (ingmarh) [inline apps/events.md:722]: Why is this needed? Or: How is this special compared to the other events?

File: apps/events.md (hunk 6 above)
Comment (ingmarh) [reply apps/events.md:253]: It doesn‘t look normal, but it is the same for me when I tried it out. I just created a bug ticket.

For now, I suggest to remove documentation for this property.

File: apps/events.md (hunk 7 above)
Comment (ingmarh) [reply apps/events.md:722]: Okay, thanks for the explanation.

**Review by ingmarh — APPROVED** [review]
